homeassistant/components/sensor/islamic_prayer_times.py

```diff
@@ -0,0 +1,235 @@
+                              update_sensors(hass, sensors, prayer_times_data),
+                              next_update_at)
+
+
+async def update_sensors(hass, sensors, prayer_times_data):
+    """Update sensors with new prayer times."""
+    # Update prayer times
+    prayer_times = prayer_times_data.get_prayer_times()
+
+    # Update all prayer times sensors
+    for sensor in sensors:
+        sensor.async_schedule_update_ha_state(True)
+
+    # Schedule next update
+    await schedule_future_update(hass, sensors, prayer_times['Midnight'],
+                                 prayer_times_data)
+
+
+class IslamicPrayerTimesData:
+    """Data object for Islamic prayer times."""
+
+    def __init__(self, latitude, longitude, calc_method):
+        """Create object to hold data."""
+        self.latitude = latitude
+        self.longitude = longitude
+        self.calc_method = calc_method
+        self.prayer_times = None
+
+    def get_prayer_times(self):
+        """Fetch prayer times for today."""
+        from prayer_times_calculator import PrayerTimesCalculator
+
+        today = datetime.today().strftime('%Y-%m-%d')
+
+        calc = PrayerTimesCalculator(latitude=self.latitude,
+                                     longitude=self.longitude,
+                                     calculation_method=self.calc_method,
+                                     date=str(today))
+
+        self.prayer_times = calc.fetch_prayer_times()
+        return self.prayer_times
+
+
+class IslamicPrayerTimeSensor(Entity):
+    """Representation of an Islamic prayer time sensor."""
+
+    ENTITY_ID_FORMAT = 'sensor.islamic_prayer_time_{}'
+
+    def __init__(self, sensor_type, prayer_times_data):
+        """Initialize the Islamic prayer time sensor."""
+        self.sensor_type = sensor_type
+        self.entity_id = self.ENTITY_ID_FORMAT.format(self.sensor_type)
+        self.prayer_times_data = prayer_times_data
+        self._display_format = "%I:%M%p"
+        self._name = self.sensor_type.capitalize()
+        prayer_time = self.prayer_times_data.prayer_times[self._name]
+        pt_dt = self.get_prayer_time_as_dt(prayer_time)
+        self._state = pt_dt.strftime(self._display_format)
+        _LOGGER.debug("%s State: %s", self.sensor_type, self._state)
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def icon(self):
+        """Icon to display in the front end."""
+        return PRAYER_TIMES_ICON
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def should_poll(self):
+        """Disable polling."""
+        return False
+
+    async def async_fire_prayer_event(self):
+        """Fire event for respective prayer time."""
+        _LOGGER.debug("Firing Event for: %s", self.sensor_type)
+        self.hass.bus.async_fire('islamic_prayer_time', {'prayer':
```

Comment:
Entities are not allowed to fire events.

Reply:
hmm okay i'll remove this.